src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
+        listSize = DEFAULT_ARRAYLIST_SIZE;
+        array = (T[])(new Object[DEFAULT_ARRAYLIST_SIZE]);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size + 1 > array.length) {
+            growArray(array.length >> 1);
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of array list range");
+        }
+        if (size + 1 > array.length) {
+            growArray(array.length >> 1);
+        }
+        System.arraycopy(array, index, array, index + 1,array.length - index - 1);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > array.length) {
+            growArray(array.length + list.size() + array.length >> 1);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of array list range");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of array list range");
+        }
+        array[index] = value;
+        if (index > size) {
+            size = index;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of array list range");
+        }
+        T result = array[index];
+        System.arraycopy(array,index + 1, array, index,array.length - index - 1);
+        size--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean isFound = false;
+        for (int i = 0; i < size; i++) {
+            if (array[i] == element || element != null && element.equals(array[i])) {
+                remove(i);
+                isFound = true;
+                break;
+            }
+        }
+        if (!isFound) {
+            throw new NoSuchElementException();
+        }
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        if (size == 0) {
+            return true;
+        }
         return false;
     }
+
+    private void growArray(int records) {
+        T[] tempArray = (T[])(new Object[array.length]);
+        System.arraycopy(array, 0, tempArray,0, array.length);
+        array = (T[])(new Object[array.length + records]);
+        System.arraycopy(tempArray, 0, array,0, array.length - records);
+    }
```

Comment:
```suggestion
    private void growArray() {
        T[] tempArray = (T[])(new Object[array.length]);
        System.arraycopy(array, 0, tempArray,0, array.length);
        array = (T[])(new Object[array.length * 1,5]);
        System.arraycopy(tempArray, 0, array,0, array.length - records);
    }
```
You have to increase the array one and a half times each time